backend.native/tests/runtime/memory/cycle_collector.kt

```diff
@@ -0,0 +1,187 @@
+    atomic.value = holder
+    return holder
+}
+
+
+fun test4() {
+    val holder = makeIt()
+    // To clean rc count coming from rememberNewContainer().
+    kotlin.native.internal.GC.collect()
+    // Request cyclic collection.
+    kotlin.native.internal.GC.collectCyclic()
+    // Ensure we processed delayed release.
+    repeat(10) {
+        // Wait a bit and process queue.
+        Worker.current.park(10)
+        Worker.current.processQueue()
+        kotlin.native.internal.GC.collect()
+    }
+    val value = @Suppress("UNCHECKED_CAST") (holder.other as? AtomicReference<Holder?>?)
+    assertTrue(value != null)
+    assertTrue(value.value == holder)
+}
+
+fun createRef(): AtomicReference<Any?> {
+    val atomic1 = AtomicReference<Any?>(null)
+    val atomic2 = AtomicReference<Any?>(null)
+    atomic1.value = atomic2
+    atomic2.value = atomic1
+    return atomic1
+}
+
+class Holder2(var value: AtomicReference<Any?>) {
+    fun switch() {
+        value = value.value as AtomicReference<Any?>
+    }
+}
+
+fun createHolder2() = Holder2(createRef())
+
+fun test5() {
+    val holder = createHolder2()
+    kotlin.native.internal.GC.collect()
+    kotlin.native.internal.GC.collectCyclic()
+    Worker.current.park(100 * 1000)
+    holder.switch()
+    kotlin.native.internal.GC.collect()
+    Worker.current.park(100 * 1000)
+    withWorker {
+        executeAfter(0L, {
+            kotlin.native.internal.GC.collect()
+        }.freeze())
+    }
+    Worker.current.park(1000)
+    assertTrue(holder.value.value != null)
+}
+
+fun test6() {
+    val atomic = AtomicReference<Any?>(null)
+    atomic.value = Pair(atomic, Holder(atomic)).freeze()
+}
+
+fun createRoot(): AtomicReference<Any?> {
+    val ref1 = AtomicReference<Any?>(null)
+    val ref2 = AtomicReference<Any?>(null)
+
+    ref1.value = Holder(ref2).freeze()
+    ref2.value = Any().freeze()
+
+    return ref1
+}
+
+fun test7() {
+    val ref1 = createRoot()
+    kotlin.native.internal.GC.collect()
+
+    kotlin.native.internal.GC.collectCyclic()
+    Worker.current.park(500 * 1000L)
+
+    Worker.start().executeAfter(0L, {}.freeze())
+    Worker.current.park(500 * 1000L)
+
+    val node = ref1.value as Holder
+    val ref2 = node.other as AtomicReference<Any?>
+    assertTrue(ref2.value != null)
+}
+
+fun array(size: Int) = Array<Any?>(size, { null })
+
+fun test8() {
+    val ref = AtomicReference<Any?>(null)
+    val obj1 = array(2)
+    val obj2 = array(1)
+    val obj3 = array(2)
+
+    obj1[0] = obj2
+    obj1[1] = obj3
+
+    obj2[0] = obj3
+
+    obj3[0] = obj2
+    obj3[1] = ref
+
+    ref.value = obj1.freeze()
+}
+
+fun createNode1(): Holder {
+    val ref = AtomicReference<Any?>(null)
+    val node2 = Holder(ref)
+    val node1 = Holder(node2)
+    ref.value = node1.freeze()
+
+    return node1
+}
+
+fun getNode2(): Holder {
+    val node1 = createNode1()
+    GC.collect()
+
+    return node1.other as Holder
+}
+
+fun test9() {
+    val worker = Worker.start()
```

Comment:
This worker isn't terminated.
IIRC, in this case leak detector is skipped during program termination, which is undesirable, considering the tests for leaks above.
Same for `test7`.